framework/src/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java

```diff
@@ -1242,15 +1242,19 @@ public AnnotationMirror createIntRangeAnnotation(long from, long to) {
     }
 
     /**
-     * Create an {@code @IntRange} annotation from the range. May return BOTTOMVAL or UNKNOWNVAL.
+     * Create an {@code @IntRange} or {@code @IntVal} annotation from the range. May return
+     * BOTTOMVAL or UNKNOWNVAL.
      */
     public AnnotationMirror createIntRangeAnnotation(Range range) {
         if (range.isNothing()) {
             return BOTTOMVAL;
         } else if (range.isEverything()) {
             return UNKNOWNVAL;
-        } else {
+        } else if (range.isWiderThan(MAX_VALUES)) {
             return createIntRangeAnnotation(range.from, range.to);
+        } else {
```

Comment:
I agree that keeping the annotation in the warning messages consistent with what the user wrote is important.  However, at this point in the code, there is no way to know if range from an explicitly written annotation or was computing by dataflow (for example `x > 5 && x <7`).  Also, the Range object could have come from an explicitly written `@IntVal` annotation which are converted to Range objects sometimes.

What if instead of using `@IntVal` when the annotation is wider than `MAX_VALUES`, `@IntVal` is used of the range is wider than 2? That way `@IntRange(from=1, to=1)` will be converted to `@IntVal(1)` and  `@IntRange(from=1, to=2)` to `@IntVal({1, 2})`.